### ISSUE TITLE
Make CI workflow resilient to multiple runs on the same sha or tag

### ISSUE THREAD

**Issue by siggy** [linkerd/linkerd2#3635]
## Background

PR #3602 guaranteed that all jobs in a CI workflow received the same copy of the repo. This made our CI workflow resilient to code changes occurring during the workflow run, both in master and the PR itself.

## Problem

Our workflow uses the sha of the repo to name global objects that are shared between jobs within a workflow. When two workflow runs occur on the same sha or tag, these two runs will collide. Specifically, in the names of the kind clusters and git repo artifacts they create.

### Collision examples

- kind cluster collision: https://github.com/linkerd/linkerd2/runs/273402477#step:5:15
- git repo artifact collision: https://github.com/linkerd/linkerd2/runs/272823368#step:2:7

## Proposal

Assign unique names to kind clusters and git repo artifacts, based on the current workflow run, in addition to the output of `bin/root-tag`.

For example, instead of:
```
TAG="$(CI_FORCE_CLEAN=1 bin/root-tag)"
export KIND_CLUSTER=github-$TAG-${{ matrix.integration_test }}`
```
do something like:
```
TAG="$(CI_FORCE_CLEAN=1 bin/root-tag)"
export KIND_CLUSTER=github-$TAG-${{ runner.tracking_id }}-${{ matrix.integration_test }}`
```

(I'm not sure if `runner.tracking_id` exists or is too long, but we need something unique like this, or something in the `{{job}}` context.) More info at https://help.github.com/en/github/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions.

#### kind cluster name

https://github.com/linkerd/linkerd2/blob/397970e917afef601103ed024caba54b27a742b9/.github/workflows/workflow.yml#L312-L313

#### git Repo artifact

https://github.com/linkerd/linkerd2/blob/397970e917afef601103ed024caba54b27a742b9/.github/workflows/workflow.yml#L78-L79

#### kubeconfig file

Note this filename is based on the kind cluster name, so fixing kind cluster naming should fix this automatically:
https://github.com/linkerd/linkerd2/blob/397970e917afef601103ed024caba54b27a742b9/.github/workflows/workflow.yml#L391

**Comment by alpeb** [comment]
I'm wondering when two runs on the same sha can occur. Is it only when an empty commit to a branch is pushed to trigger the build, when the branch still has a build running?

**Comment by ihcsim** [comment]
> I'm wondering when two runs on the same sha can occur.

Like a new tag and the latest master HEAD? Or do we handle that differently now in `workflow.yaml`?

**Comment by siggy** [comment]
@alpeb @ihcsim Good points. I've updated the issue title and description to better describe the situation. We don't exactly rely on the sha, but rather on the output of `bin/root-tag`, which is either a sha, or a tag (if one exists).

The situations where this can occur include:
- master is updated to a commit, then soon after a tag is set to that same commit
- the repo is tagged at a commit, then soon after the same tag is updated to a different commit

**Comment by alpeb** [comment]
Turns out the job and runner context are pretty limited:
```json
job:
{
  "status": "Success"
}

runner:
{
  "os": "Linux",
  "tool_cache": "/opt/hostedtoolcache",
  "temp": "/home/runner/work/_temp",
  "workspace": "/home/runner/work/linkerd2"
}
```
and I couldn't find anything useful in the `github` context either.

I'm gonna try generating a random number in bash and storing it in the workflow's temp dir. Will have to check that dir is not shared across runs.

**Comment by siggy** [comment]
@alpeb I could be wrong, but I wonder if the GitHub Actions UI is obfuscating additional fields from the `runner` and `job` contexts? I ask because I can see env vars like `RUNNER_TRACKING_ID` that are presumably mapped from a context?

https://github.com/linkerd/linkerd2/runs/273480582#step:2:50

**Comment by alpeb** [comment]
I tested this:
```yaml
    - name: Dump runner.tracking_id
      env:
        RUNNER_ID: ${{ runner.tracking_id }}
      run: echo "$RUNNER_ID"
    - name: Check if runner.tracking_id is empty
      env:
        RUNNER_ID: ${{ runner.tracking_id == ''}}
      run: echo "$RUNNER_ID"
    - name: Dump runner.tracking.id
      env:
        RUNNER_ID: ${{ runner.tracking.id }}
      run: echo "$RUNNER_ID"
    - name: Check if runner.tracking.id is empty
      env:
        RUNNER_ID: ${{ runner.tracking.id == ''}}
      run: echo "$RUNNER_ID"
``
which showed `runner.tracking_id` and `runner.tracking.id` don't exist.
I'll give `RUNNER_TRACKING_ID` a try.

**Comment by alpeb** [comment]
Turns out not all environment variables are carried over when using the `container` directive. For example here I'm dumping the env vars in the `go_unit_tests` step, and `RUNNER_TRACKING_ID` and friends are missing:
https://github.com/alpeb/linkerd2/runs/275474709#step:3:25

**Comment by alpeb** [comment]
> I'm gonna try generating a random number in bash and storing it in the workflow's temp dir. Will have to check that dir is not shared across runs.

Answering myself :-P I had forgotten that each job uses a brand new environment and so the temp dir starts anew for each job.

OTHO, I stumbled upon [this](https://github.community/t5/GitHub-Actions/Check-Suite-identifier-as-an-environment-variable/m-p/35498) that describes a way to retrieve the "check suite ID" through github's API and `jq`. That's only relevant for a single run, but we still can use the full API response to actually see info about all the runs for a given SHA. I'm not sure yet how to select from that list the current run. If that ends up not being possible, we could at the very least check if there's a single run with status running, and fail the build instructing the user to first cancel the other build that is running for this SHA...

**Comment by siggy** [comment]
I just noticed that `INVOCATION_ID` and `RUNNER_TRACKING_ID` change between jobs within a single workflow, so those are not helpful.

**Comment by siggy** [comment]
@alpeb Thanks for all the digging. That [check suite](https://github.community/t5/GitHub-Actions/Check-Suite-identifier-as-an-environment-variable/m-p/35498) approach is interesting, though I think it grabs the latest `check_suite` item, which could change between workflow jobs.

I tested with:
```bash
curl -s -H "Accept: application/vnd.github.antiope-preview+json" "https://api.github.com/repos/linkerd/linkerd2/commits/4ea87eae0dfbbad33698155a75b1842573020321/check-suites"
```

I'm reluctant to just fail the build. Let's hold off on doing this until there's a way to uniquely identify a workflow run (or it becomes painful enough to warrant failing the build).

**Comment by alpeb** [comment]
> I think it grabs the latest `check_suite` item, which could change between workflow jobs.

I think I don't understand your objection. From what I've tested, that API will return an array where each element corresponds to each github actions build run that has been triggered for the supplied sha. Each element describes the status (running, completed), the conclusion (success, failure) and (from the looks of it) an id that is unique for that build run.

I'm ok with holding off on this one, just wanted to clarify we're on the same page.

**Comment by alpeb** [comment]
This was partially addressed in #4001, and will be finished in #4033

**Comment by alpeb** [comment]
Fixed in #4033!
